{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import os\n",
    "import module\n",
    "from Bio import pairwise2\n",
    "import ali_count\n",
    "from multiprocessing import Pool\n",
    "import rmsdm\n",
    "import math\n",
    "import knee\n",
    "import copy\n",
    "from sklearn.cluster import OPTICS\n",
    "from sklearn.cluster import DBSCAN\n",
    "import hdbscan\n",
    "import ntpath\n",
    "from sklearn.metrics import silhouette_score\n",
    "from scipy.spatial import distance\n",
    "from tqdm import tqdm \n",
    "from tabulate import tabulate\n",
    "\n",
    "class Amino_acid:\n",
    "    def __init__(self, name, coords, protein_name, number):\n",
    "        self.coords = coords\n",
    "        self.name = name\n",
    "        self.protein_name = protein_name\n",
    "        self.number = number\n",
    "        self.calculate_start_end()\n",
    "    \n",
    "    def print_warning_lack_of_atom(self, *atoms):\n",
    "        atoms_lack = [atom for atom in atoms if atom not in self.coords]\n",
    "        if len(atoms_lack) != 0:\n",
    "            print(\"Warning: Lack of atom(s) {} in file {} for amino acid {}\".format(atoms_lack, self.protein_name, self.number)) \n",
    "        \n",
    "    def calculate_start_end(self):\n",
    "        if self.name == 'ALA':\n",
    "            self.start = self.coords.get('CA', None)\n",
    "            self.end = self.coords.get('CB', None)\n",
    "            self.print_warning_lack_of_atom('CA', 'CB')\n",
    "        elif self.name == 'ARG':\n",
    "            self.start = self.coords.get('CD', None)\n",
    "            if 'NH1' in self.coords and 'NH2' in self.coords:\n",
    "                self.end = (self.coords['NH1'] + self.coords['NH2']) / 2\n",
    "            else: \n",
    "                self.end = None\n",
    "            self.print_warning_lack_of_atom('CD', 'NH1', 'NH2')\n",
    "        elif self.name == 'ASN':\n",
    "            self.start = self.coords.get('CB', None)\n",
    "            if 'OD1' in self.coords and 'ND2' in self.coords:\n",
    "                self.end = (self.coords['OD1'] + self.coords['ND2']) / 2\n",
    "            else:\n",
    "                self.end = None\n",
    "            self.print_warning_lack_of_atom('CB', 'OD1', 'ND2')\n",
    "        elif self.name == 'ASP':\n",
    "            self.start = self.coords.get('CB', None)\n",
    "            if 'OD1' in self.coords and 'OD2' in self.coords:\n",
    "                self.end = (self.coords['OD1'] + self.coords['OD2']) / 2 \n",
    "            else:\n",
    "                self.end = None\n",
    "            self.print_warning_lack_of_atom('CB', 'OD1', 'OD2')\n",
    "        elif self.name == 'CYS':\n",
    "            self.start = self.coords.get('CA', None)\n",
    "            self.end = self.coords.get('SG', None)\n",
    "            self.print_warning_lack_of_atom('CA', 'SG')\n",
    "        elif self.name == 'GLN':\n",
    "            self.start = self.coords.get('CG', None)\n",
    "            if 'NE2' in self.coords and 'OE1' in self.coords:\n",
    "                self.end = (self.coords['NE2'] + self.coords['OE1']) / 2 \n",
    "            else:\n",
    "                self.end = None\n",
    "            self.print_warning_lack_of_atom('CG', 'NE2', 'OE1')\n",
    "        elif self.name == 'GLU':\n",
    "            self.start = self.coords.get('CG', None)\n",
    "            if 'OE1' in self.coords and 'OE2' in self.coords:\n",
    "                self.end = (self.coords['OE1'] + self.coords['OE2']) / 2 \n",
    "            else:\n",
    "                self.end = None\n",
    "            self.print_warning_lack_of_atom('CG', 'OE2', 'OE2')\n",
    "        elif self.name == 'GLY':\n",
    "            self.start = self.coords.get('CA', None)\n",
    "            if 'N' in self.coords and 'C' in self.coords:\n",
    "                self.end = (self.coords['N'] + self.coords['C']) / 2 \n",
    "            else:\n",
    "                self.end = None\n",
    "            self.print_warning_lack_of_atom('CA', 'N', 'C')\n",
    "        elif self.name == 'HIS':\n",
    "            self.start =  self.coords.get('CG', None)\n",
    "            if 'CE1' in self.coords and 'NE2' in self.coords:\n",
    "                self.end = (self.coords['CE1'] + self.coords['NE2']) / 2 \n",
    "            else:\n",
    "                self.end = None\n",
    "            self.print_warning_lack_of_atom('CG', 'CE1', 'NE2')\n",
    "        elif self.name == 'ILE':\n",
    "            self.start = self.coords.get('CB', None)\n",
    "            self.end = self.coords.get('CD1', None)\n",
    "            self.print_warning_lack_of_atom('CB', 'CD1')\n",
    "        elif self.name == 'LEU':\n",
    "            self.start = self.coords.get('CB', None)\n",
    "            if 'CD1' in self.coords and 'CD2' in self.coords:\n",
    "                self.end = (self.coords['CD1'] + self.coords['CD2']) / 2 \n",
    "            else:\n",
    "                self.end = None\n",
    "            self.print_warning_lack_of_atom('CB', 'CD1', 'CD2')\n",
    "        elif self.name == 'LYS':\n",
    "            self.start =  self.coords.get('CG', None)\n",
    "            self.end =  self.coords.get('NZ', None)\n",
    "            self.print_warning_lack_of_atom('CG', 'NZ')\n",
    "        elif self.name == 'MET':\n",
    "            self.start = self.coords.get('CA', None)\n",
    "            self.end = self.coords.get('SD', None)\n",
    "            self.print_warning_lack_of_atom('CA', 'SD')\n",
    "        elif self.name == 'PHE':\n",
    "            self.start = self.coords.get('CG', None)\n",
    "            self.end = self.coords.get('CZ', None)\n",
    "            self.print_warning_lack_of_atom('CG', 'CZ')\n",
    "        elif self.name == 'PRO':\n",
    "            self.start = self.coords.get('CA', None) \n",
    "            if 'CG' in self.coords and 'CD' in self.coords:\n",
    "                self.end = (self.coords['CG'] + self.coords['CD']) / 2 \n",
    "            else:\n",
    "                self.start = None\n",
    "            self.print_warning_lack_of_atom('CG', 'CA', 'CD')\n",
    "        elif self.name == 'SER':\n",
    "            self.start = self.coords.get('CA', None) \n",
    "            self.end = self.coords.get('OG', None)\n",
    "            self.print_warning_lack_of_atom('CA', 'OG')\n",
    "        elif self.name == 'THR':\n",
    "            self.start = self.coords.get('CA', None) \n",
    "            if 'CG2' in self.coords and 'OG1' in self.coords:\n",
    "                self.end = (self.coords['CG2'] + self.coords['OG1']) / 2 \n",
    "            else:\n",
    "                self.end = None\n",
    "            self.print_warning_lack_of_atom('CA', 'CG2', 'OG1')\n",
    "        elif self.name == 'TRP':\n",
    "            self.start = self.coords.get('CD1', None)\n",
    "            if 'CZ2' in self.coords and 'CZ3' in self.coords:\n",
    "                self.end = (self.coords['CZ2'] + self.coords['CZ3']) / 2 \n",
    "            else:\n",
    "                self.end = None\n",
    "            self.print_warning_lack_of_atom('CD1', 'CZ2', 'CZ3')\n",
    "        elif self.name == 'TYR':\n",
    "            self.start = self.coords.get('CG', None)\n",
    "            self.end = self.coords.get('CZ', None)\n",
    "            self.print_warning_lack_of_atom('CG', 'CZ')\n",
    "        elif self.name == 'VAL': \n",
    "            self.start = self.coords.get('CA', None)\n",
    "            if 'CG1' in self.coords and 'CG2' in self.coords:\n",
    "                self.end = (self.coords['CG1'] + self.coords['CG2']) / 2 \n",
    "            else:\n",
    "                self.end = None\n",
    "            self.print_warning_lack_of_atom('CA', 'CG1', 'CG2')\n",
    "        else:\n",
    "            self.start = None\n",
    "            self.end = None\n",
    "\n",
    "class PDB:\n",
    "    def __init__(self):\n",
    "        self.amino_acid_num_to_name = {}\n",
    "        self.amino_acid_num_to_atom_name_to_coords = {}\n",
    "        self.seq = []\n",
    "    \n",
    "    def glue(self, seq):\n",
    "        glue_seq = []\n",
    "        for i, val in enumerate(seq):\n",
    "            if i == 0: glue_seq.append(val)\n",
    "            else:\n",
    "                if val == seq[i-1]: pass\n",
    "                else: glue_seq.append(val)\n",
    "        return glue_seq\n",
    "    \n",
    "    def name_of_at(self, line):\n",
    "        return(line[12:16].replace(' ',''))\n",
    "\n",
    "    def num_of_aa(self, line):\n",
    "        return(int(line[22:26]))  \n",
    "\n",
    "    def name_of_ch(self, line):\n",
    "        return line[21:22] \n",
    "    \n",
    "    def name_of_aa(self, line):\n",
    "        return line[16:20] .replace(' ','')       \n",
    "\n",
    "    def num_of_at(self, line):\n",
    "        return line[6:11].replace(' ','') \n",
    "    \n",
    "    def reading(self, filename):\n",
    "        self.filename = os.path.abspath(filename)\n",
    "        self.base_filename = ntpath.basename(os.path.abspath(filename))\n",
    "        chain = 'No chain'\n",
    "        with open(filename, 'r') as f:\n",
    "            lines = f.readlines()\n",
    "            for line in lines:\n",
    "                if line.startswith(\"ATOM\"):\n",
    "                    if chain != self.name_of_ch(line) and chain != 'No chain':\n",
    "                        raise Exception('Exception: One pdb-file should contain one chain. File {} contains more, than one chain'.format(filename))\n",
    "                    \n",
    "                    chain = self.name_of_ch(line)\n",
    "                    self.amino_acid_num_to_name[self.num_of_aa(line)] = self.name_of_aa(line)\n",
    "                    self.seq.append(self.num_of_aa(line))\n",
    "                    \n",
    "                    if self.num_of_aa(line) in self.amino_acid_num_to_atom_name_to_coords:\n",
    "                        self.amino_acid_num_to_atom_name_to_coords[self.num_of_aa(line)][self.name_of_at(line)] = np.array([float(line[30:38]), float(line[38:46]), float(line[46:54])])\n",
    "                    else:\n",
    "                        self.amino_acid_num_to_atom_name_to_coords[self.num_of_aa(line)] = {}\n",
    "                        self.amino_acid_num_to_atom_name_to_coords[self.num_of_aa(line)][self.name_of_at(line)] = np.array([float(line[30:38]), float(line[38:46]), float(line[46:54])])\n",
    "                else:\n",
    "                    pass\n",
    "        self.seq = self.glue(self.seq)    \n",
    "        \n",
    "        self.amino_acids = {}\n",
    "        for amino_acid_num in self.amino_acid_num_to_name:\n",
    "            self.amino_acids[amino_acid_num] = Amino_acid(self.amino_acid_num_to_name[amino_acid_num], self.amino_acid_num_to_atom_name_to_coords[amino_acid_num], self.base_filename, amino_acid_num)\n",
    "\n",
    "class PDBS_alignment:\n",
    "    def __init__(self, dir_with_aligned_pdbs, filename_fasta):\n",
    "        self.dir_with_aligned_pdbs = dir_with_aligned_pdbs\n",
    "        self.filename_fasta = filename_fasta\n",
    "        self.molecules = []         \n",
    "        self.alignment_fasta = []\n",
    "        self.alignment = []\n",
    "        self.dirs_in_fasta_alignment_and_in_pdb_folder = []   \n",
    "        \n",
    "        self.get_fasta_alignment()\n",
    "        self.check_if_alignment_is_ok()\n",
    "    \n",
    "    def check_if_alignment_is_ok(self):\n",
    "        list_of_bad_backbone_aa = {}\n",
    "        for i in range(len(self.alignment)):\n",
    "            for col in range(len(self.alignment[0])):\n",
    "                if self.alignment[i][col] != '-':\n",
    "                    if all(k in self.molecules[i].amino_acid_num_to_atom_name_to_coords[self.alignment[i][col]] for k in ('N', 'CA', 'C', 'O')):\n",
    "                        continue\n",
    "                    else:\n",
    "                        if i in list_of_bad_backbone_aa:\n",
    "                            list_of_bad_backbone_aa[i].append(self.alignment[i][col])\n",
    "                        else:\n",
    "                            list_of_bad_backbone_aa[i] = [self.alignment[i][col]]\n",
    "        for i in list_of_bad_backbone_aa:\n",
    "            print('Warning: PDB entry {} contains incomplete set of backbone atoms for amino acid(s)'.format(self.dirs_in_fasta_alignment_and_in_pdb_folder[i]), list_of_bad_backbone_aa[i])\n",
    "        \n",
    "        for i in range(len(self.alignment)):\n",
    "            for j in range(i, len(self.alignment)):\n",
    "                rmsds = []\n",
    "                for col in range(len(self.alignment[0])):\n",
    "                    if self.alignment[i][col] == '-' or self.alignment[j][col] == '-':\n",
    "                        continue\n",
    "                   \n",
    "                    if 'CA' in self.molecules[i].amino_acid_num_to_atom_name_to_coords[self.alignment[i][col]] and 'CA' in self.molecules[j].amino_acid_num_to_atom_name_to_coords[self.alignment[j][col]]:\n",
    "                        P1 = []\n",
    "                        P2 = []\n",
    "                        P1.append(self.molecules[i].amino_acid_num_to_atom_name_to_coords[self.alignment[i][col]]['CA'])\n",
    "                        P2.append(self.molecules[j].amino_acid_num_to_atom_name_to_coords[self.alignment[j][col]]['CA'])\n",
    "                        rmsds.append(rmsdm.rmsd(P1, P2))\n",
    "                if len(rmsds) != 0 :\n",
    "                    mean_rmsd = sum(rmsds) / len(rmsds)\n",
    "                    if mean_rmsd > 9:\n",
    "                         print('Warning: Average RMSD between {} and {} is {} what may indicate poor alignment quality of the two proteins'.format(self.dirs_in_fasta_alignment_and_in_pdb_folder[i], self.dirs_in_fasta_alignment_and_in_pdb_folder[j],str(mean_rmsd)))\n",
    "    \n",
    "    def get_conserv_positions(self, perc_of_gaps=0.05, max_content_of_mismatch=0.05, mismatch_threshold=None):\n",
    "        \n",
    "        def align_analysis_rec(heap, align, cut, RMSD_MATR):\n",
    "            i_max = np.unravel_index(RMSD_MATR.argmax(), RMSD_MATR.shape)[0]\n",
    "            j_max = np.unravel_index(RMSD_MATR.argmax(), RMSD_MATR.shape)[1]\n",
    "            max = np.max(RMSD_MATR)\n",
    "            if max < cut: \n",
    "                return\n",
    "            else: \n",
    "                max1 = RMSD_MATR[i_max].sum()\n",
    "                max2 = RMSD_MATR[j_max].sum()\n",
    "            if max1 > max2:\n",
    "                align[i_max][heap] = '-'\n",
    "                RMSD_MATR[i_max][:] = 0\n",
    "                RMSD_MATR[:][i_max] = 0\n",
    "            else: \n",
    "                align[j_max][heap] = '-' \n",
    "                RMSD_MATR[j_max][:] = 0\n",
    "                RMSD_MATR[:][j_max] = 0\n",
    "            align_analysis_rec(heap, align, cut, RMSD_MATR)\n",
    "        \n",
    "        def RMSD_MATR_one_column(heap):\n",
    "            RMSD_MATR = np.zeros((len(self.alignment),len(self.alignment))) \n",
    "            for i in range(len(self.alignment)):\n",
    "                for j in range(i, len(self.alignment)): \n",
    "                    if self.alignment[i][heap] == '-' or self.alignment[j][heap] == '-': \n",
    "                        RMSD_MATR[i][j] = 0\n",
    "                        RMSD_MATR[j][i] = 0\n",
    "                    else:\n",
    "                        if all(k in self.molecules[i].amino_acid_num_to_atom_name_to_coords[self.alignment[i][heap]] for k in ('N', 'CA', 'C', 'O')) and all(k in self.molecules[j].amino_acid_num_to_atom_name_to_coords[self.alignment[j][heap]] for k in ('N', 'CA', 'C', 'O')):\n",
    "                            P1 = []\n",
    "                            P2 = []\n",
    "                            P1.append(self.molecules[i].amino_acid_num_to_atom_name_to_coords[self.alignment[i][heap]]['N'])\n",
    "                            P1.append(self.molecules[i].amino_acid_num_to_atom_name_to_coords[self.alignment[i][heap]]['CA'])\n",
    "                            P1.append(self.molecules[i].amino_acid_num_to_atom_name_to_coords[self.alignment[i][heap]]['C'])\n",
    "                            P1.append(self.molecules[i].amino_acid_num_to_atom_name_to_coords[self.alignment[i][heap]]['O'])\n",
    "                            P2.append(self.molecules[j].amino_acid_num_to_atom_name_to_coords[self.alignment[j][heap]]['N'])\n",
    "                            P2.append(self.molecules[j].amino_acid_num_to_atom_name_to_coords[self.alignment[j][heap]]['CA'])\n",
    "                            P2.append(self.molecules[j].amino_acid_num_to_atom_name_to_coords[self.alignment[j][heap]]['C'])\n",
    "                            P2.append(self.molecules[j].amino_acid_num_to_atom_name_to_coords[self.alignment[j][heap]]['O'])\n",
    "                            RMSD_MATR[i][j] = rmsdm.rmsd(P1, P2)\n",
    "                            RMSD_MATR[j][i] = RMSD_MATR[i][j]\n",
    "                        else:\n",
    "                            RMSD_MATR[i][j] = 0\n",
    "                            RMSD_MATR[j][i] = 0   \n",
    "            return RMSD_MATR\n",
    "        \n",
    "        def align_analysis(heap, align, cut):\n",
    "            RMSD_MATR = RMSD_MATR_one_column(heap)\n",
    "            align_analysis_rec(heap, align, cut, RMSD_MATR) \n",
    "        \n",
    "        def get_max_rmsd_in_column(heap):\n",
    "            RMSD_MATR = RMSD_MATR_one_column(heap)   \n",
    "            max = np.max(RMSD_MATR)\n",
    "            return max\n",
    "        \n",
    "        def num_of_let(alignment, col): \n",
    "            pos_with_letters = []\n",
    "            for j in range(len(alignment)):\n",
    "                if alignment[j][col] != '-': \n",
    "                    pos_with_letters.append(j)\n",
    "            return len(pos_with_letters)\n",
    "        \n",
    "        def get_elbow(rmsds):\n",
    "            y_axis = sorted(rmsds)\n",
    "            Perc_5 = math.floor(len(y_axis) * 0.05)\n",
    "            x_axis = [i for i in range(len(rmsds))]\n",
    "            data = np.asarray(list(zip(x_axis[Perc_5:len(x_axis) - Perc_5],y_axis[Perc_5:len(x_axis) - Perc_5])))\n",
    "            elbow_index = knee.find_elbow(data, knee.get_data_radiant(data))\n",
    "            cut = data[elbow_index][1] \n",
    "            if cut == 0 or cut == y_axis[Perc_5]:\n",
    "                cut = y_axis[-1]\n",
    "            print('Info: The automatically selected cut-off value to discriminate spatially aligned from misaligned residues is {} A'.format(round(cut,3)))\n",
    "            return cut\n",
    "        \n",
    "        if mismatch_threshold is None:\n",
    "            heaps = []\n",
    "            for col in range(len(self.alignment[0])):\n",
    "                if num_of_let(self.alignment, col) >= len(self.alignment) * (1 - perс_of_gaps):\n",
    "                    heaps.append(col)\n",
    "            assert not len(heaps) == 0, \"It seems that the number of common core residues is not sufficient for cluster analysis. You should increase the 'max_content_of_gaps' parameter or reconsider your input alignment\"\n",
    "            rmsds = [get_max_rmsd_in_column(heap) for heap in heaps]\n",
    "            mismatch_threshold = get_elbow(rmsds)\n",
    "        align_tmp = copy.deepcopy(self.alignment)\n",
    "        heaps = []\n",
    "        for col in range(len(self.alignment[0])):\n",
    "            if num_of_let(self.alignment, col) > len(self.alignment) * (1 - max_content_of_mismatch):\n",
    "                heaps.append(col)\n",
    "        for heap in heaps:\n",
    "            align_analysis(heap, align_tmp, mismatch_threshold)\n",
    "        heaps = []\n",
    "        for col in range(len(align_tmp[0])):\n",
    "            if num_of_let(align_tmp, col) > len(align_tmp) * (1 - max_content_of_mismatch):\n",
    "                heaps.append(col)\n",
    "        return heaps\n",
    "    \n",
    "    def get_fasta_alignment(self):\n",
    "        def convertion(a):    \n",
    "            aminoac = {\n",
    "            \"G\" : 'GLY',\n",
    "            \"L\" : 'LEU',\n",
    "            \"Y\" : 'TYR',\n",
    "            \"S\" : 'SER',\n",
    "            \"E\" : 'GLU',\n",
    "            \"Q\" : 'GLN',\n",
    "            \"D\" : 'ASP',\n",
    "            \"N\" : 'ASN',\n",
    "            \"F\" : 'PHE',\n",
    "            \"A\" : 'ALA',\n",
    "            \"K\" : 'LYS',\n",
    "            \"R\" : 'ARG',\n",
    "            \"H\" : 'HIS',\n",
    "            \"C\" : 'CYS',\n",
    "            \"V\" : 'VAL',\n",
    "            \"P\" : 'PRO',\n",
    "            \"W\" : 'TRP',\n",
    "            \"I\" : 'ILE',\n",
    "            \"M\" : 'MET',\n",
    "            \"T\" : 'THR',\n",
    "            \"GLY\" : 'G',\n",
    "            \"LEU\" : 'L',\n",
    "            \"TYR\" : 'Y',\n",
    "            \"SER\" : 'S',\n",
    "            \"GLU\" : 'E',\n",
    "            \"GLN\" : 'Q',\n",
    "            \"ASP\" : 'D',\n",
    "            \"ASN\" : 'N',\n",
    "            \"PHE\" : 'F',\n",
    "            \"ALA\" : 'A',\n",
    "            \"LYS\" : 'K',\n",
    "            \"ARG\" : 'R',\n",
    "            \"HIS\" : 'H',\n",
    "            \"CYS\" : 'C',\n",
    "            \"VAL\" : 'V',\n",
    "            \"PRO\" : 'P',\n",
    "            \"HYP\" : 'hP',\n",
    "            \"TRP\" : 'W',\n",
    "            \"ILE\" : 'I',\n",
    "            \"MET\" : 'M',\n",
    "            \"THR\" : 'T',\n",
    "            }\n",
    "            if a in aminoac:\n",
    "                b = aminoac[a]\n",
    "            else:\n",
    "                b = 'X'\n",
    "            return b\n",
    "    \n",
    "        with open(self.filename_fasta, 'r') as fasta:\n",
    "            names_of_prot_in_fasta = module.names_of_prot(fasta)\n",
    "            seq_from_fasta = module.seq_from_fasta(fasta)\n",
    "        dirs_in_pdb_folder = os.listdir(self.dir_with_aligned_pdbs)\n",
    "        dict_fasta_to_dir = {}\n",
    "        dict_dir_to_fasta = {}\n",
    "\n",
    "        assert len(names_of_prot_in_fasta) == len(dirs_in_pdb_folder), 'Error: number of proteins in fasta-file is not equal to number of proteins in pdb-files folder'\n",
    "        for i, name_in_fasta in enumerate(names_of_prot_in_fasta):\n",
    "            if name_in_fasta + '.pdb' in dirs_in_pdb_folder:\n",
    "                dir = name_in_fasta + '.pdb'\n",
    "                dict_fasta_to_dir[name_in_fasta] = name_in_fasta + '.pdb'\n",
    "                dict_dir_to_fasta[name_in_fasta + '.pdb'] = name_in_fasta\n",
    "            elif name_in_fasta.replace(':','_')+'.pdb' in dirs_in_pdb_folder:\n",
    "                dir = name_in_fasta.replace(':', '_') + '.pdb'\n",
    "                dict_fasta_to_dir[name_in_fasta] = name_in_fasta.replace(':', '_') + '.pdb'\n",
    "                dict_dir_to_fasta[name_in_fasta.replace(':', '_') + '.pdb'] = name_in_fasta\n",
    "            elif name_in_fasta.split(':')[0]+'.pdb' in dirs_in_pdb_folder:\n",
    "                dir = name_in_fasta.split(':')[0]+'.pdb'\n",
    "                dict_fasta_to_dir[name_in_fasta] = name_in_fasta.split(':')[0]+'.pdb'\n",
    "                dict_dir_to_fasta[name_in_fasta.split(':')[0]+'.pdb'] = name_in_fasta\n",
    "            else:\n",
    "                raise Exception('Warning: Lack of file ' + names_of_prot_in_fasta[i] + '.pdb in input folder')\n",
    "            self.alignment.append([])\n",
    "            self.dirs_in_fasta_alignment_and_in_pdb_folder.append(dir)\n",
    "            self.alignment_fasta.append(seq_from_fasta[i])\n",
    "            mol = PDB()\n",
    "            mol.reading(os.path.join(self.dir_with_aligned_pdbs, dir))\n",
    "            self.molecules.append(mol)  \n",
    "            seq_from_pdb =''\n",
    "            for j in mol.seq:\n",
    "                seq_from_pdb += convertion(mol.amino_acid_num_to_name[j])\n",
    "            alignment_tmp = pairwise2.align.globalxx(seq_from_fasta[i].replace(\"-\", \"\"), seq_from_pdb)\n",
    "            c = [i for i in range(len(seq_from_fasta[i].replace(\"-\", \"\")))]\n",
    "            c_res = ali_count.fasta_bind_site(c, alignment_tmp[0][0], alignment_tmp[0][1])\n",
    "            k = 0\n",
    "            for j in seq_from_fasta[i]:\n",
    "                if j == '-':\n",
    "                    self.alignment[-1].append('-')\n",
    "                    continue\n",
    "                elif j != '-':\n",
    "                    if c_res[k] != -1:\n",
    "                        self.alignment[-1].append(mol.seq[c_res[k]])\n",
    "                    elif c_res[k] == -1:\n",
    "                        self.alignment[-1].append('-')\n",
    "                    k += 1\n",
    "      \n",
    "     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "aligned_fasta = './mustguseal/strcore_A-8sbasog07j9spd/strcore/strcore_A-8sbasog07j9spd.fasta'\n",
    "aligned_pdb_folder = './mustguseal/strcore_A-8sbasog07j9spd/strcore/aligned_pdbs'\n",
    "output = 'out'\n",
    "method = 'optics'\n",
    "min_cluster_size = 5\n",
    "min_samples = 5\n",
    "eps = 5\n",
    "perc_of_gaps = 0.05\n",
    "max_content_of_mismatch = 0.05\n",
    "v = 1.0\n",
    "mismatch_threshold = 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Lack of atom(s) ['CG2'] in file 264_6l8q_G.pdb for amino acid 86\n",
      "Warning: Lack of atom(s) ['NZ'] in file 264_6l8q_G.pdb for amino acid 386\n",
      "Warning: Lack of atom(s) ['CG2'] in file 264_6l8q_G.pdb for amino acid 751\n",
      "Warning: Lack of atom(s) ['CD', 'NH1', 'NH2'] in file 268_2gbf_A.pdb for amino acid 38\n",
      "Warning: Lack of atom(s) ['CD', 'NH1', 'NH2'] in file 268_2gbf_A.pdb for amino acid 39\n",
      "Warning: Lack of atom(s) ['CG2', 'OG1'] in file 268_2gbf_A.pdb for amino acid 40\n",
      "Warning: Lack of atom(s) ['CG2', 'OG1'] in file 268_2gbf_A.pdb for amino acid 42\n",
      "Warning: Lack of atom(s) ['CD1', 'CD2'] in file 268_2gbf_A.pdb for amino acid 43\n",
      "Warning: Lack of atom(s) ['CG', 'CZ'] in file 268_2gbf_A.pdb for amino acid 96\n",
      "Warning: Lack of atom(s) ['OD1', 'ND2'] in file 268_2gbf_A.pdb for amino acid 442\n",
      "Warning: Lack of atom(s) ['CG', 'NZ'] in file 268_2gbf_A.pdb for amino acid 503\n",
      "Warning: Lack of atom(s) ['CG', 'NZ'] in file 319_4wjl_B.pdb for amino acid 97\n",
      "Warning: Lack of atom(s) ['CB', 'CD1'] in file 319_4wjl_B.pdb for amino acid 413\n",
      "Warning: Lack of atom(s) ['CG', 'NZ'] in file 319_4wjl_B.pdb for amino acid 553\n",
      "Warning: Lack of atom(s) ['NZ'] in file 319_4wjl_B.pdb for amino acid 669\n",
      "Warning: Lack of atom(s) ['CD', 'NH1', 'NH2'] in file 332_2ecf_A.pdb for amino acid 61\n",
      "Warning: Lack of atom(s) ['CG', 'NZ'] in file 332_2ecf_A.pdb for amino acid 137\n",
      "Warning: Lack of atom(s) ['CG', 'NE2', 'OE1'] in file 332_2ecf_A.pdb for amino acid 138\n",
      "Warning: Lack of atom(s) ['CG', 'OE2', 'OE2'] in file 332_2ecf_A.pdb for amino acid 139\n",
      "Warning: Lack of atom(s) ['CG', 'NZ'] in file 332_2ecf_A.pdb for amino acid 141\n",
      "Warning: Lack of atom(s) ['CG', 'NZ'] in file 332_2ecf_A.pdb for amino acid 368\n",
      "Warning: Lack of atom(s) ['CG', 'NZ'] in file 332_2ecf_A.pdb for amino acid 370\n",
      "Warning: Lack of atom(s) ['CG', 'NZ'] in file 337_2d5l_A.pdb for amino acid 76\n",
      "Warning: Lack of atom(s) ['CG', 'NE2', 'OE1'] in file 337_2d5l_A.pdb for amino acid 126\n",
      "Warning: Lack of atom(s) ['CG', 'OE2', 'OE2'] in file 337_2d5l_A.pdb for amino acid 179\n",
      "Warning: Lack of atom(s) ['CG', 'NZ'] in file 337_2d5l_A.pdb for amino acid 470\n",
      "Warning: Lack of atom(s) ['CD', 'NH1', 'NH2'] in file 337_2d5l_A.pdb for amino acid 531\n",
      "Warning: Lack of atom(s) ['SD'] in file 342_4q1v_B.pdb for amino acid 77\n",
      "Warning: Lack of atom(s) ['SD'] in file 342_4q1v_B.pdb for amino acid 112\n",
      "Warning: Lack of atom(s) ['SD'] in file 342_4q1v_B.pdb for amino acid 205\n",
      "Warning: Lack of atom(s) ['SD'] in file 342_4q1v_B.pdb for amino acid 209\n",
      "Warning: Lack of atom(s) ['SD'] in file 342_4q1v_B.pdb for amino acid 232\n",
      "Warning: Lack of atom(s) ['SD'] in file 342_4q1v_B.pdb for amino acid 235\n",
      "Warning: Lack of atom(s) ['SD'] in file 342_4q1v_B.pdb for amino acid 302\n",
      "Warning: Lack of atom(s) ['SD'] in file 342_4q1v_B.pdb for amino acid 309\n",
      "Warning: Lack of atom(s) ['SD'] in file 342_4q1v_B.pdb for amino acid 352\n",
      "Warning: Lack of atom(s) ['OE2', 'OE2'] in file 342_4q1v_B.pdb for amino acid 470\n",
      "Warning: Lack of atom(s) ['SD'] in file 342_4q1v_B.pdb for amino acid 478\n",
      "Warning: Lack of atom(s) ['SD'] in file 342_4q1v_B.pdb for amino acid 500\n",
      "Warning: Lack of atom(s) ['SD'] in file 342_4q1v_B.pdb for amino acid 541\n",
      "Warning: Lack of atom(s) ['SD'] in file 342_4q1v_B.pdb for amino acid 548\n",
      "Warning: Lack of atom(s) ['SD'] in file 342_4q1v_B.pdb for amino acid 608\n",
      "Warning: Lack of atom(s) ['SD'] in file 342_4q1v_B.pdb for amino acid 613\n",
      "Warning: Lack of atom(s) ['SD'] in file 342_4q1v_B.pdb for amino acid 638\n",
      "Warning: Lack of atom(s) ['SD'] in file 342_4q1v_B.pdb for amino acid 644\n",
      "Warning: Lack of atom(s) ['NZ'] in file 342_4q1v_B.pdb for amino acid 667\n",
      "Warning: Lack of atom(s) ['SD'] in file 342_4q1v_B.pdb for amino acid 690\n",
      "Warning: PDB entry 264_6l8q_G.pdb contains incomplete set of backbone atoms for amino acid(s) [177]\n"
     ]
    }
   ],
   "source": [
    "alignment = PDBS_alignment(aligned_pdb_folder, aligned_fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "heaps = alignment.get_conserv_positions(perc_of_gaps=perc_of_gaps, max_content_of_mismatch=max_content_of_mismatch, mismatch_threshold=mismatch_threshold)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "class One_column_clustering:\n",
    "    def __init__(self, col, alignment, method='hdbscan'):\n",
    "        self.col = col\n",
    "        self.alignment = alignment\n",
    "        self.method = method\n",
    "        \n",
    "    def clustering(self, min_cluster_size=5, min_samples=3, eps=1):\n",
    "        clust_matr = []\n",
    "        self.mols_and_aa = []\n",
    "        for i in range(len(alignment.alignment)):\n",
    "            num_of_aa = alignment.alignment[i][self.col]\n",
    "            if num_of_aa != '-' and alignment.molecules[i].amino_acids[num_of_aa].start is not None and alignment.molecules[i].amino_acids[num_of_aa].end is not None:\n",
    "                self.mols_and_aa.append((alignment.molecules[i], num_of_aa)) \n",
    "                clust_matr.append(np.hstack((alignment.molecules[i].amino_acids[num_of_aa].start, alignment.molecules[i].amino_acids[num_of_aa].end)))\n",
    "        clust_matr = np.array(clust_matr)\n",
    "        if self.method == 'optics':\n",
    "            clusterer = OPTICS(metric='euclidean', n_jobs=-1, min_samples=min_samples)\n",
    "        elif self.method == 'hdbscan':\n",
    "            hdbscan.HDBSCAN(metric='euclidean', min_cluster_size=min_cluster_size, min_samples=min_samples) \n",
    "        elif self.method == 'dbscan':\n",
    "            clusterer = DBSCAN(metric='euclidean', n_jobs=-1, eps=eps, min_samples=min_samples)\n",
    "        db = clusterer.fit(clust_matr)\n",
    "        self.lab = db.labels_\n",
    "        if list(self.lab).count(-1) == 0 and len(set(self.lab)) == 2 or list(self.lab).count(-1) == 1 and len(set(self.lab)) == 3:\n",
    "            self.sil = silhouette_score(clust_matr[self.lab != -1], self.lab[self.lab != -1], metric='euclidean')\n",
    "            dist_matr = np.array([[distance.euclidean(clust_matr[i], clust_matr[j]) for i in range(len(clust_matr))] for j in range(len(clust_matr))])\n",
    "            mean_diams_clusters = [dist_matr[self.lab == i].T[self.lab == j].mean() for i in set(self.lab) for j in set(self.lab) if i != j and i != -1 and j != -1] \n",
    "            self.diam = max(mean_diams_clusters) \n",
    "            self.score = self.sil * self.diam\n",
    "        else: \n",
    "            self.sil = None\n",
    "            self.diam = None\n",
    "            self.score = None\n",
    "            \n",
    "    def printing_py_result(self, output_folder):\n",
    "        for i in range(len(self.lab)):\n",
    "            if self.lab[i] == 0:\n",
    "                f.write('cmd.color(\"lightpink\", \"{} and resi {}\")\\n'.format(self.mols_and_aa[i][0].base_filename, self.mols_and_aa[i][1]))\n",
    "            elif self.lab[i] == 1:\n",
    "                f.write('cmd.color(\"palegreen\", \"{} and resi {}\")\\n'.format(self.mols_and_aa[i][0].base_filename, self.mols_and_aa[i][1]))\n",
    "            elif self.lab[i] == 2:\n",
    "                f.write('cmd.color(\"lightblue\", \"{} and resi {}\")\\n'.format(self.mols_and_aa[i][0].base_filename, self.mols_and_aa[i][1]))\n",
    "            elif self.lab[i] == 3:\n",
    "                f.write('cmd.color(\"paleyellow\", \"{} and resi {}\")\\n'.format(self.mols_and_aa[i][0].base_filename, self.mols_and_aa[i][1]))\n",
    "            elif self.lab[i] == 4:\n",
    "                f.write('cmd.color(\"wheat\", \"{} and resi {}\")\\n'.format(self.mols_and_aa[i][0].base_filename, self.mols_and_aa[i][1]))\n",
    "            elif self.lab[i] == 5:\n",
    "                f.write('cmd.color(\"palecyan\", \"{} and resi {}\")\\n'.format(self.mols_and_aa[i][0].base_filename, self.mols_and_aa[i][1]))\n",
    "            elif self.lab[i] == 6:\n",
    "                f.write('cmd.color(\"lightorange\", \"{} and resi {}\")\\n'.format(self.mols_and_aa[i][0].base_filename, self.mols_and_aa[i][1]))\n",
    "            elif self.lab[i] == 7:\n",
    "                f.write('cmd.color(\"bluewhite\", \"{} and resi {}\")\\n'.format(self.mols_and_aa[i][0].base_filename, self.mols_and_aa[i][1]))\n",
    "            elif self.lab[i] == -1:\n",
    "                f.write('cmd.color(\"{}\", \"{} and resi {}\")\\n'.format(self.lab[i] + 3, self.mols_and_aa[i][0].base_filename, self.mols_and_aa[i][1]))\n",
    "            else:\n",
    "                f.write('cmd.color(\"{}\", \"{} and resi {}\")\\n'.format(self.lab[i] + 3, self.mols_and_aa[i][0].base_filename, self.mols_and_aa[i][1]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "cls = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 16%|████████████▊                                                                    | 54/342 [01:57<10:24,  2.17s/it]\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-84-b16d64753c94>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mheap\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mtqdm\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mheaps\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      2\u001b[0m     \u001b[0mcl\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mOne_column_clustering\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mheap\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0malignment\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mmethod\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 3\u001b[1;33m     \u001b[0mcl\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mclustering\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmin_cluster_size\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mmin_cluster_size\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mmin_samples\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mmin_samples\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0meps\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0meps\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      4\u001b[0m     \u001b[0mcls\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcl\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-82-4fcebfdbe710>\u001b[0m in \u001b[0;36mclustering\u001b[1;34m(self, min_cluster_size, min_samples, eps)\u001b[0m\n\u001b[0;32m     20\u001b[0m         \u001b[1;32melif\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmethod\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;34m'dbscan'\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     21\u001b[0m             \u001b[0mclusterer\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mDBSCAN\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmetric\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'euclidean'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mn_jobs\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;33m-\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0meps\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0meps\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mmin_samples\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mmin_samples\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 22\u001b[1;33m         \u001b[0mdb\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mclusterer\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mclust_matr\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     23\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mlab\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mdb\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mlabels_\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     24\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mlist\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mlab\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcount\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m-\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;36m0\u001b[0m \u001b[1;32mand\u001b[0m \u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mset\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mlab\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;36m2\u001b[0m \u001b[1;32mor\u001b[0m \u001b[0mlist\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mlab\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcount\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m-\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;36m1\u001b[0m \u001b[1;32mand\u001b[0m \u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mset\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mlab\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;36m3\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mD:\\Anaconda3\\lib\\site-packages\\sklearn\\cluster\\_optics.py\u001b[0m in \u001b[0;36mfit\u001b[1;34m(self, X, y)\u001b[0m\n\u001b[0;32m    257\u001b[0m              \u001b[0mleaf_size\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mleaf_size\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mmetric\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmetric\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    258\u001b[0m              \u001b[0mmetric_params\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmetric_params\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mp\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mp\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mn_jobs\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mn_jobs\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 259\u001b[1;33m              max_eps=self.max_eps)\n\u001b[0m\u001b[0;32m    260\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    261\u001b[0m         \u001b[1;31m# Extract clusters from the calculated orders and reachability\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mD:\\Anaconda3\\lib\\site-packages\\sklearn\\utils\\validation.py\u001b[0m in \u001b[0;36minner_f\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m     70\u001b[0m                           FutureWarning)\n\u001b[0;32m     71\u001b[0m         \u001b[0mkwargs\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m{\u001b[0m\u001b[0mk\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0marg\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mk\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0marg\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mzip\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0msig\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mparameters\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0margs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m}\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 72\u001b[1;33m         \u001b[1;32mreturn\u001b[0m \u001b[0mf\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     73\u001b[0m     \u001b[1;32mreturn\u001b[0m \u001b[0minner_f\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     74\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mD:\\Anaconda3\\lib\\site-packages\\sklearn\\cluster\\_optics.py\u001b[0m in \u001b[0;36mcompute_optics_graph\u001b[1;34m(X, min_samples, max_eps, metric, p, metric_params, algorithm, leaf_size, n_jobs)\u001b[0m\n\u001b[0;32m    497\u001b[0m                             \u001b[0mprocessed\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mprocessed\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mX\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mX\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mnbrs\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mnbrs\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    498\u001b[0m                             \u001b[0mmetric\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mmetric\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mmetric_params\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mmetric_params\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 499\u001b[1;33m                             p=p, max_eps=max_eps)\n\u001b[0m\u001b[0;32m    500\u001b[0m     \u001b[1;32mif\u001b[0m \u001b[0mnp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mall\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mnp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0misinf\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mreachability_\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    501\u001b[0m         warnings.warn(\"All reachability values are inf. Set a larger\"\n",
      "\u001b[1;32mD:\\Anaconda3\\lib\\site-packages\\sklearn\\cluster\\_optics.py\u001b[0m in \u001b[0;36m_set_reach_dist\u001b[1;34m(core_distances_, reachability_, predecessor_, point_index, processed, X, nbrs, metric, metric_params, p, max_eps)\u001b[0m\n\u001b[0;32m    513\u001b[0m     \u001b[1;31m# we may be doing some work twice.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    514\u001b[0m     indices = nbrs.radius_neighbors(P, radius=max_eps,\n\u001b[1;32m--> 515\u001b[1;33m                                     return_distance=False)[0]\n\u001b[0m\u001b[0;32m    516\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    517\u001b[0m     \u001b[1;31m# Getting indices of neighbors that have not been processed\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mD:\\Anaconda3\\lib\\site-packages\\sklearn\\neighbors\\_base.py\u001b[0m in \u001b[0;36mradius_neighbors\u001b[1;34m(self, X, radius, return_distance, sort_results)\u001b[0m\n\u001b[0;32m    970\u001b[0m                               sort_results=sort_results)\n\u001b[0;32m    971\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 972\u001b[1;33m                 \u001b[1;32mfor\u001b[0m \u001b[0ms\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mgen_even_slices\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mX\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mn_jobs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    973\u001b[0m             )\n\u001b[0;32m    974\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0mreturn_distance\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mD:\\Anaconda3\\lib\\site-packages\\joblib\\parallel.py\u001b[0m in \u001b[0;36m__call__\u001b[1;34m(self, iterable)\u001b[0m\n\u001b[0;32m   1069\u001b[0m                 \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_backend\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mstop_call\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1070\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_managed_backend\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1071\u001b[1;33m                 \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_terminate_backend\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1072\u001b[0m             \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_jobs\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1073\u001b[0m             \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_pickle_cache\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;32mNone\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mD:\\Anaconda3\\lib\\site-packages\\joblib\\parallel.py\u001b[0m in \u001b[0;36m_terminate_backend\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    762\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m_terminate_backend\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    763\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_backend\u001b[0m \u001b[1;32mis\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[1;32mNone\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 764\u001b[1;33m             \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_backend\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mterminate\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    765\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    766\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m_dispatch\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mbatch\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mD:\\Anaconda3\\lib\\site-packages\\joblib\\_parallel_backends.py\u001b[0m in \u001b[0;36mterminate\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    241\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_pool\u001b[0m \u001b[1;32mis\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[1;32mNone\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    242\u001b[0m             \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_pool\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 243\u001b[1;33m             \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_pool\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mterminate\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m  \u001b[1;31m# terminate does a join()\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    244\u001b[0m             \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_pool\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;32mNone\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    245\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mD:\\Anaconda3\\lib\\multiprocessing\\pool.py\u001b[0m in \u001b[0;36mterminate\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    546\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_state\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mTERMINATE\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    547\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_worker_handler\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_state\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mTERMINATE\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 548\u001b[1;33m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_terminate\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    549\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    550\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0mjoin\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mD:\\Anaconda3\\lib\\multiprocessing\\util.py\u001b[0m in \u001b[0;36m__call__\u001b[1;34m(self, wr, _finalizer_registry, sub_debug, getpid)\u001b[0m\n\u001b[0;32m    222\u001b[0m                 sub_debug('finalizer calling %s with args %s and kwargs %s',\n\u001b[0;32m    223\u001b[0m                           self._callback, self._args, self._kwargs)\n\u001b[1;32m--> 224\u001b[1;33m                 \u001b[0mres\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_callback\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m*\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_args\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_kwargs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    225\u001b[0m             \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_weakref\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_callback\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_args\u001b[0m \u001b[1;33m=\u001b[0m\u001b[0;31m \u001b[0m\u001b[0;31m\\\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    226\u001b[0m                             \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_kwargs\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_key\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;32mNone\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mD:\\Anaconda3\\lib\\multiprocessing\\pool.py\u001b[0m in \u001b[0;36m_terminate_pool\u001b[1;34m(cls, taskqueue, inqueue, outqueue, pool, worker_handler, task_handler, result_handler, cache)\u001b[0m\n\u001b[0;32m    592\u001b[0m         \u001b[0mutil\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mdebug\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'joining worker handler'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    593\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mthreading\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcurrent_thread\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mis\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[0mworker_handler\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 594\u001b[1;33m             \u001b[0mworker_handler\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    595\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    596\u001b[0m         \u001b[1;31m# Terminate workers which haven't already finished.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mD:\\Anaconda3\\lib\\threading.py\u001b[0m in \u001b[0;36mjoin\u001b[1;34m(self, timeout)\u001b[0m\n\u001b[0;32m   1042\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1043\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mtimeout\u001b[0m \u001b[1;32mis\u001b[0m \u001b[1;32mNone\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1044\u001b[1;33m             \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_wait_for_tstate_lock\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1045\u001b[0m         \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1046\u001b[0m             \u001b[1;31m# the behavior of a negative timeout isn't documented, but\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mD:\\Anaconda3\\lib\\threading.py\u001b[0m in \u001b[0;36m_wait_for_tstate_lock\u001b[1;34m(self, block, timeout)\u001b[0m\n\u001b[0;32m   1058\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mlock\u001b[0m \u001b[1;32mis\u001b[0m \u001b[1;32mNone\u001b[0m\u001b[1;33m:\u001b[0m  \u001b[1;31m# already determined that the C code is done\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1059\u001b[0m             \u001b[1;32massert\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_is_stopped\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1060\u001b[1;33m         \u001b[1;32melif\u001b[0m \u001b[0mlock\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0macquire\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mblock\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1061\u001b[0m             \u001b[0mlock\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mrelease\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1062\u001b[0m             \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_stop\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "for heap in tqdm(heaps):\n",
    "    cl = One_column_clustering(heap, alignment, method)\n",
    "    cls.append(cl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = open(os.path.join('out', 'RESULT.py'), 'w')\n",
    "f.write('cmd.bg_color(\"white\")\\n')\n",
    "for i in range(len(alignment.molecules)):\n",
    "    f.write('cmd.load(r\"{}\", \"{}\")\\n'.format(alignment.molecules[i].filename, alignment.molecules[i].base_filename))\n",
    "    f.write('cmd.color(\"silver\", \"{}\")\\n'.format(alignment.molecules[i].base_filename))        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "cls_sorted = sorted(cls, key=lambda x: x.score if x.score is not None else -9999, reverse=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "for cl in cls_sorted[0:10]:\n",
    "    if cl.score is not None:\n",
    "        cl.printing_py_result(output)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "table = [[\"Rank\", \"Silhouette_score\", \"Diameter\", \"Score\"]]\n",
    "for i, cl in enumerate(cls_sorted[0:10]):\n",
    "    if cl.score is not None:\n",
    "        table.append([i + 1, round(cl.sil, 3), round(cl.diam, 3), round(cl.score, 3)])\n",
    "\n",
    "with open(os.path.join(output, 'RESULTS.txt'), 'w') as f:\n",
    "    f.write(tabulate(table)+\"\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
